app/assets/javascripts/latlong_map.coffee

```diff
@@ -3,38 +3,67 @@ marker = null
 # fallback to PVTA if we can't geocode the incident's location
 PVTA = { lat: 42.105552, lng: -72.596511 }
 
-createMap = (mapSelector, latLng, placeInitialMarker) ->
+createMap = (mapSelector, latLng, newValues) ->
   map = new google.maps.Map mapSelector[0], { zoom: 15, tilt: 0, center: latLng }
-  if placeInitialMarker
-    placeMarker map, latLng
+  placeMarker map, latLng, newValues
   google.maps.event.addListener map, 'click', (event) ->
-    placeMarker map, event.latLng
-    fillLatLngFields event.latLng
+    latLong = { lat: event.latLng.lat(), lng: event.latLng.lng() }
+    placeMarker map, latLong, true
+
 
 fillLatLngFields = (latLng) ->
-  $('#incident_latitude').val Number(latLng.lat().toFixed(5))
-  $('#incident_longitude').val Number(latLng.lng().toFixed(5))
+  $('#incident_latitude').val Number(latLng.lat.toFixed(5))
+  $('#incident_longitude').val Number(latLng.lng.toFixed(5))
 
 initLatLngMap = (mapSelector) ->
-  if mapSelector.data 'lat'
+  if mapSelector[0].dataset.lat
     lat = parseFloat mapSelector.data('lat')
     lng = parseFloat mapSelector.data('lng')
-    createMap mapSelector, { lat: lat, lng: lng }, true
+    createMap mapSelector, { lat: lat, lng: lng }, false
   else
     location = mapSelector.data 'location'
     geocoder = new google.maps.Geocoder()
     geocoder.geocode address: location, (results) ->
       if results.length > 0
         result = results[0]
-        createMap mapSelector, result.geometry.location
+        createMap mapSelector, result.geometry.location, false
       else
         $('.no-geocode-alert').slideDown()
-        createMap mapSelector, PVTA
+        createMap mapSelector, PVTA, false
 
-placeMarker = (map, latLng) ->
+placeMarker = (map, latLng, newValues) ->
   unless marker == null
     marker.setMap null
   marker = new google.maps.Marker position: latLng, map: map
+  fillLatLngFields latLng
+  if newValues
+    reverseGeocode latLng
+
+reverseGeocode = (latLng) ->
+  geocoder = new google.maps.Geocoder()
+  geocoder.geocode location: latLng, (results, status) ->
+    address = results[0].address_components
+    [street, zip, town, state] = ["", undefined, undefined, undefined]
+    address.forEach (part) ->
+      if part.types.includes("street_number")
+        street += part.short_name + " "
+      if part.types.includes("route")
+        street += part.short_name
```

Comment:
Are the `address_components` guaranteed to be in this order? I just wonder if it would be possible to get something like `"Main Street 123 "`

Reply:
I think in our area of use the address components are "guaranteed" to be in this order if they exist. The only examples I have found where this is not the case is in different countries where address components are recorded in different orders.

Reply:
Alright, I'm willing to risk it, [but...](https://developers.google.com/maps/documentation/geocoding/overview#GeocodingResponses)

> Note the following facts about the address_components[] array:
>
>  * ...
>  * The format of the response is not guaranteed to remain the same between requests. In particular, the number of address_components varies based on the address requested and can change over time for the same address. **A component can change position in the array**. The type of the component can change. A particular component may be missing in a later response.